### PR TITLE
New CUDA compiler

### PR DESCRIPTION
**Pull request by linehill**
This patch introduces a new CUDA compiler - `cucc`. This is meant to replace `cuspvc` and aims to be a drop-in replacement for nvcc. It is still a work in progress but manages to build ~300 CUDA benchmarks successfully in HeCBench.

There is also a way to sneak the `cucc` into CMake for compiling CUDA projects without having `find_package(CUDA)` or `find_package(CUDAToolkit)`. See docs/Using.md for instructions.

### CODE REVIEW

File: tests/cuda/CMakeLists.txt [hunk 1]
```diff
@@ -10,16 +10,68 @@ function(add_compile_test)
...
+  ENVIRONMENT CUCC_VERBOSE=1
+  PASS_REGULAR_EXPRESSION "Executing: .*hipcc .*")
+
+# Check -Xcompiler passes options to the driven compiler.
+add_compile_test(NAME cucc-xcompiler-opt
+  COMPILE_OPTIONS
+  "-Xcompiler -DFOO=123 -Xcompiler=-DBAR=321 check-macros.cu -E -o /dev/null")
+
+# Check --version response can be user defined. A use case is for
+# masquerading another compiler.
```
Comment (pjaaskel) [inline tests/cuda/CMakeLists.txt:66]:
```suggestion
# maximal compatibility when there are too strict version checks in the build files.
```

File: tests/cuda/some.c [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#ifdef __cpluspluc
```
Comment (pjaaskel) [inline tests/cuda/some.c:1]:
```suggestion
#ifdef __cplusplus
```

File: tests/post-install/CMake-CUDA/.#CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+linehill@fiji.23322:1694497632
```
Comment (pjaaskel) [inline tests/post-install/CMake-CUDA/.#CMakeLists.txt:1]: Accidentally added file?

**Review by pjaaskel — CHANGES_REQUESTED** [review]
Looks promising!

File: bin/cucc.in [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+import os
+import subprocess
+import sys
+from shlex import quote
+
+# If true, we change the behavior of this tool so that the CMake
+# considers this tool to be NVvidia's nvcc. This allows us to compile
+# CUDA code in CMake without find_package().
+# TODO: Add an option or env-var for controlling the value of this.
+MASQUERADE_NVCC = True
```
[the remark this answers is not in the document]
Comment (linehill) [reply bin/cucc.in]: Done.

File: bin/cucc.in [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+    Here we define options we want to capture and act upon - for
+    example, translate NVCC specific options to corresponding one in
+    hipcc, filter out options, raise errors or warnings on unsupported
+    options, etc.
+
+    Unknown arguments are passed to hipcc.
+    """
+    parser = argparse.ArgumentParser(
+        description=("CUDA compiler for chipStar and a drop-in replacement"
```
[the remark this answers is not in the document]
Comment (linehill) [reply bin/cucc.in]: Done.

File: bin/cucc.in [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+
+    if os.environ.get('CUCC_VERBOSE') == "1":
+        print(f"Executing: {' '.join(map(quote, hipcc_args))}",
+              file=sys.stderr)
+
+    result = subprocess.run(hipcc_args, check=False, shell=False)
+    sys.exit(result.returncode)
+
+if __name__ == '__main__':
+    warn("cucc is work-in-progess."
```
[the remark this answers is not in the document]
Comment (linehill) [reply bin/cucc.in]: Done.

File: docs/Using.md [hunk 7]
```diff
@@ -40,7 +40,78 @@ Note that long-running GPU compute kernels can trigger hang detection mechanism
 
 ### Compiling CUDA application directly with chipStar
 
-Compilation of CUDA sources without changing the sources, can be done in two ways. The first is to replace calls of the nvcc compiler with calls of the wrapper script <CHIP-install-path>/bin/cuspv in Makefiles. The wrapper script will call clang with the correct flags. The other way is possible when using CMake: use `find_package(HIP REQUIRED CONFIG)` and then use `target_link_libraries(<YOUR_TARGET> hip::device)`. However the project must be compiled with Clang (a version supported by HIP). Note that it's not necessary to have Nvidia's CUDA installed.
+CUDA sources can be compiled directly without converting the sources
+to HIP first with our `cucc` CUDA compiler which can be used as
```
[the remark this answers is not in the document]
Comment (linehill) [reply docs/Using.md]: Done except left out the "script". It's unnecessary implementation detail to me.

File: docs/Using.md [hunk 8]
```diff
@@ -40,7 +40,78 @@ Note that long-running GPU compute kernels can trigger hang detection mechanism
...
+drop-in replacement for
+[nvcc](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html).
+Beware though - the tool, and our CUDA runtime headers, are still a
```
[the remark this answers is not in the document]
Comment (linehill) [reply docs/Using.md]: Done.

File: docs/Using.md [hunk 9]
```diff
@@ -40,7 +40,78 @@ Note that long-running GPU compute kernels can trigger hang detection mechanism
...
+work in progress and some features are missing not not supported. Here
+is non-exhaustive list of known issues and limitations:
```
[the remark this answers is not in the document]
Comment (linehill) [reply docs/Using.md]: Done.

File: docs/Using.md [hunk 10]
```diff
@@ -40,7 +40,78 @@ Note that long-running GPU compute kernels can trigger hang detection mechanism
...
+* Inline assembly snippets within device code are not supported.
+
+* CUDA sources without any includes do not compile for now.
+  Workaround: include <cuda_runtime.h>.
+
+* Some nvcc preprocessor defines are not set currently -
+  e.g. `__CUDA_ARCH__` and `__CUDA_ARCH_LIST__`.
+
+* Sources depending on CUDA libraries (thrust, cuBLAS, cuFFT, etc)
+  won't compile for now with possibly obscure error messages.
```
[the remark this answers is not in the document]
Comment (linehill) [reply docs/Using.md]: Done.

File: docs/Using.md [hunk 11]
```diff
@@ -40,7 +40,78 @@ Note that long-running GPU compute kernels can trigger hang detection mechanism
...
+
+* Warp functions ending with `_sync` are not currently
+  supported. Workaround: call the legacy warp functions instead if
+  possible - the current requirement is that all threads in the warp
```
[the remark this answers is not in the document]
Comment (linehill) [reply docs/Using.md]: Done.

File: docs/Using.md [hunk 12]
```diff
@@ -40,7 +40,78 @@ Note that long-running GPU compute kernels can trigger hang detection mechanism
...
+  must converge at the same warp function.
+
+* Some runtime API types and functions might be undeclared.
+
+#### Direct CUDA Usage with CMake
```
[the remark this answers is not in the document]
Comment (linehill) [reply docs/Using.md]: Done.

File: docs/Using.md [hunk 13]
```diff
@@ -40,7 +40,78 @@ Note that long-running GPU compute kernels can trigger hang detection mechanism
...
+
+(experimental; currently the support has been tested with CMake 3.24.3)
+
+CUDA projects using CMake can be targeted at chipStar. However, this
```
[the remark this answers is not in the document]
Comment (linehill) [reply docs/Using.md]: Done.

File: docs/Using.md [hunk 14]
```diff
@@ -40,7 +40,78 @@ Note that long-running GPU compute kernels can trigger hang detection mechanism
...
+Create dummy `libcudart.so`, `libcudart_static.a` and `libcudadevrt.so`
+files and place them under the chipStar library install directory:
+
+```
+touch <chipStar-install-dir>/lib/libcudart.so
+touch <chipStar-install-dir>/lib/libcudart_static.a
+touch <chipStar-install-dir>/lib/libcudadevrt.so
+```
+
+And invoke cmake as followed:
```
[the remark this answers is not in the document]
Comment (linehill) [reply docs/Using.md]: Done.

File: include/hip/spirv_fixups.h [hunk 15]
```diff
@@ -49,11 +49,28 @@
 
 #endif // __HIP_DEVICE_COMPILE__
 
+
+#if defined(__CHIPSTAR_MASQUERADE_CUDA__) && defined(__HIP__)
```
[the remark this answers is not in the document]
Comment (linehill) [reply include/hip/spirv_fixups.h]: Done.

File: tests/cuda/some.c (hunk 2 above)
Comment (linehill) [reply tests/cuda/some.c]: Done.

File: tests/post-install/CMake-CUDA/.#CMakeLists.txt (hunk 3 above)
Comment (linehill) [reply tests/post-install/CMake-CUDA/.#CMakeLists.txt]: Yep. Should be removed now.

File: include/cuspv/cuda_runtime.h [hunk 16]
```diff
@@ -31,8 +31,17 @@
 #ifndef __CUDA_RUNTIME_H__
```
[the remark this answers is not in the document]
Comment (linehill) [reply include/cuspv/cuda_runtime.h:37]: Added a check.

File: include/cuspv/cuda_runtime.h [hunk 17]
```diff
@@ -26,11 +26,21 @@
...
 
+// Catch accidental inclusion of NVidia's cuda_runtime.h header.
+#if defined(__CUDA_RUNTIME_H__) && !defined(__CHIP_CUDA_RUNTIME_H__)
+#error                                                                         \
+    "NVidia's cuda_runtime.h appears to be included (unsupported by chipStar)"
+#endif
+
 // Using same include guard name as in CUDA SDK's cuda_runtime.h so it
 // gets excluded if it happens to be in the include search paths.
 #ifndef __CUDA_RUNTIME_H__
```
Comment (pjaaskel) [inline include/cuspv/cuda_runtime.h:37]: Does this now fail if we include cuspv's cuda_runtime.h twice? Shouldn't this define moved inside the __CHIP_CUDA_RUNTIME_H__ guard?

**Review by pjaaskel — CHANGES_REQUESTED** [review]

File: include/cuspv/cuda_runtime.h (hunk 17 above)
Comment (linehill) [reply include/cuspv/cuda_runtime.h:37]: > Does this now fail if we include cuspv's cuda_runtime.h twice?

It's designed to not fail when included multiple times. I added a test for this just now.

**Pull request closed without merging**